ansible/roles/k8s_install/tasks/main.yml

```
@@ -0,0 +1,160 @@
+    extensions:
+      - 'yaml'
+      - 'yml'
+  with_items: "{{ ansible_inventory_sources }}"
+  tags:
+    - always
+
+- name: "Assert cluster_name is not set to cluster.local"
+  ansible.builtin.assert:
+    that:
+      - "hostvars[ groups['k8s_cluster'][0] ]['cluster_name'] not in ['cluster.local']"
+  tags:
+    - always
+
+- name: "Display base configuration"
+  ansible.builtin.debug:
+    msg: "{{ item.msg }}: {{ item.value }}"
+  with_items:
+    - { msg: "Pre Install Hook", value: "{{ pre_install_hook }}" }
+    - { msg: "Kubernetes Product", value: "{{ kubeprovider.name }}" }
+    - { msg: "Kubernetes Installer Path", value: "{{ kubeprovider.path }}" }
+    - { msg: "Kubernetes Cluster Name", value: "{{ hostvars[ groups['k8s_cluster'][0] ]['cluster_name'] }}" }
+    - { msg: "Kubectl Config (default)", value: "{{ kubeconfig_file }}" }
+  tags:
+    - always
+
+- name: "Check supported kubernetes installer"
+  ansible.builtin.assert:
+    that:
+      - "kubeprovider.name in supported_kube_installer"
+    fail_msg: "Requested kubernetes installer not supported"
+    quiet: True
+  tags:
+    - always
+
+- name: "Setting facts"
+  ansible.builtin.set_fact:
+    cluster_name: "{{ hostvars[ groups['k8s_cluster'][0] ]['cluster_name'] }}"
+  tags:
+    - always
+
+
+###
+### Installation section
+###
+
+### (Kubespray) Install K8s cluster
+- name: "Execute kubespray installer"
+  ansible.builtin.import_tasks: kubespray_install.yml
+  tags:
+    - install
+  when:
+    - "kubeprovider.name == 'kubespray'"
+    - "kube_install_mode == 'install'"
+
+### Install K8s client after cluster install
+- name: "Ensure k8s client configuration"
+  ansible.builtin.import_tasks: k8s_client.yml
+  tags:
+    - always
+  when:
+    - "kube_install_mode == 'install' or kube_install_mode == 'upgrade'"
+
+
+###
+### Upgrade section
+###
+
+### Retrieve k8s server versions
+- name: "Determine kubernetes server version"
+  kubernetes.core.k8s_cluster_info:
+  register: _kubernetes_server_info
+  failed_when: _kubernetes_server_info.version.server.kubernetes.gitVersion is not defined
+  tags:
+    - always
+
+- name: "Set current and desired version facts"
+  ansible.builtin.set_fact:
+    kube_desired_version: "{{ kube_version |regex_findall('[0-9]+') |join }}"
+    #kube_current_version: "{{ _kubernetes_server_info.stdout |from_json |json_query('serverVersion.gitVersion') |regex_findall('[0-9]+') |join }}" #kubectl output
+    kube_current_version: "{{ _kubernetes_server_info.version.server.kubernetes.gitVersion |regex_findall('[0-9]+') |join }}"
+  tags:
+    - always
+
+### (Kubespray) Upgrade
+- debug:
+    msg: "Upgrade skipped as kubernetes version {{ kube_desired_version }} <= {{ kube_current_version }}"
+  when:
+    - "kube_desired_version <= kube_current_version"
+
+- name: "Check kubespray upgrade"
+  ansible.builtin.import_tasks: kubespray_upgrade.yml
+  tags:
+    - upgrade
+  when:
+    - "kubeprovider.name == 'kubespray'"
```

just a suggestion to normalize the input to make sure the condition always works.

```suggestion
    - "(kubeprovider.name | lower) == 'kubespray'"
```